src/app/utils/xml-utils.ts

```diff
@@ -36,3 +36,141 @@ export function replaceNotWordChar(textContent: string) {
 export function removeSpaces(textContent: string) {
   return textContent.replace(/\s/g, '');
 }
+
+/**
+ * It removes excessive spaces, any tabulation, new lines and non-word characters
+ * @param textContent string
+ * @returns string
+ */
+export function normalizeSpaces(textContent: string) {
+  return textContent.replace(/[\s]{2,}|\n|\t|\r/g, ' ').trimStart().trimEnd();
+}
+
+/**
+* Significant text sometimes is split inside two or more text evt-element inside the main one, especially when it contains new line characters.
+* This function returns a string with all the text element chained
+* @param n XMLElement
+* @returns string
+*/
+export function chainFirstChildTexts(elem: XMLElement): string {
+  if (elem === undefined) { return ''};
+  const evtTextElements = {
+    '#text': 'nodeValue',
+    'p': 'textContent',
+  };
+  const evtTextComplexElements = ['choice', 'app', 'l', 'quote', 'p', 'lg'];
+  let out = '';
+  elem.childNodes.forEach((x) => (evtTextElements[x.nodeName] !== undefined) ? out += x[ evtTextElements[x.nodeName] ] : (
+    evtTextComplexElements.includes(x.nodeName) ? out += chainDeepTexts(x) : '' ))
+
+  return out;
+}
+
+/**
+ * Retrieve and chain textContent of all descendents
+ * @param elem ChildNode
+ * @returns out string
+ */
+export function chainDeepTexts(elem: ChildNode): string {
+  const evtInnerTextElements = ['#text', 'reg', 'corr', 'rdg'];
+  const textProperty = 'textContent';
+  let out = '';
+  elem.childNodes.forEach((x) => (evtInnerTextElements.includes(x.nodeName)) ? ((x[textProperty] !== null) ? out += x[textProperty] : '') : '')
+
+  return out;
+}
+
+/**
+* Retrieve external elements searching between provided elements types and filtering on attributes base
+* It searches all document for a bibl with the correct xml:id.
+* It would be faster if we knew the id or unique element to search in
+* @param element XMLElement
+* @param attrSourceNames
+* @param attrTargetName
+* @returns array of XMLElement
+*/
+export function getExternalElements(elem: XMLElement, attrSourceNames: string[], attrTargetName: string, elTypes: string): XMLElement[] {
+  const sourceIDs = attrSourceNames.map((x) => elem.getAttribute(x));
+  const sourcesToFind = sourceIDs.filter((x) => x).map((x) => x.replace('#',''));
+
+  if (sourcesToFind.length === 0) {
+    return [];
+  }
+
+  return Array.from(elem.ownerDocument.querySelectorAll<XMLElement>(elTypes))
+    .filter((x) => sourcesToFind.includes(x.getAttribute(attrTargetName)))
+}
+
+export function isAnalogue(elem: XMLElement, markerAttrs: string[]): boolean {
+  return (markerAttrs.includes(elem.getAttribute('type')));
+}
+
+export function isSource(elem: XMLElement, attrs: string[]): boolean {
+  let validAttrs = false;
+  attrs.forEach((attr) => { if (elem.getAttribute(attr) !== null) { validAttrs = true } });
+
+  return (validAttrs);
+}
+
```

Comment:
`isAnalogue` and `isSource` are not generic xml utils, but are specific function strictly connected to analogue and source. Moreover I don't understand how you validate the fact that an element is an analogue or source. Can you explain so that I can try and help you deciding how to edit this part of code

Reply:
Analogues are elements with a signature attributes type (es: "parallelPassage") set by the user in edition_config.json 
Sources are elements with one of the attributes (es: "source" or "target") set by the user in edition_config.json